### PR TITLE
[InstCombine] Optimise x / sqrt(y / z) with fast-math pattern.

### LINKED ISSUE #76605: clang is suboptimal for `x / sqrt(y / z)` with fast-math
Test case:
https://godbolt.org/z/joGE9G4KM
```c
#include<math.h>

double test_sqrt1(double x, double y, double z) { 
  return x / sqrt(y / z); 
}

double test_sqrt2(double x, double y, double z) { 
  return x * sqrt(z / y); 
}
```

My expectation is that the output of `test_sqrt1` will be the same as `test_sqrt2`. The same transformatin is true not only for `sqrt` but also for `cbrt`.

### PR DESCRIPTION
**Pull request by zjaffal**
Replace the pattern with
x * sqrt(z/y)

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp [hunk 1]
```diff
@@ -1701,6 +1701,31 @@ static Instruction *foldFDivPowDivisor(BinaryOperator &I,
...
+
+  Value *Y, *Z;
+  auto *DivOp = dyn_cast<Instruction>(II->getOperand(0));
+  if (!DivOp || !DivOp->hasOneUse() || !DivOp->hasAllowReassoc() ||
+      !I.hasAllowReciprocal())
+    return nullptr;
+  if (match(DivOp, m_FDiv(m_Value(Y), m_Value(Z)))) {
+    Value *SwapDiv = Builder.CreateFDivFMF(Z, Y, DivOp);
+    Value *NewSqrt = Builder.CreateIntrinsic(II->getIntrinsicID(),
+                                             II->getType(), {SwapDiv}, II);
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1723]:
```suggestion
    Value *NewSqrt = Builder.CreateUnaryIntrinsic(Intrinsic::sqrt, SwapDiv, II);
```

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp [hunk 2]
```diff
@@ -1701,6 +1701,31 @@ static Instruction *foldFDivPowDivisor(BinaryOperator &I,
...
 
+/// Convert div to mul if we have an sqrt divisor iff sqrt's operand is a fdiv
+/// instruction.
+static Instruction *foldFDivSqrtDivisor(BinaryOperator &I,
+                                        InstCombiner::BuilderTy &Builder) {
+  // X / sqrt(Y / Z) -->  X * sqrt(Z / Y)
+  Value *Op0 = I.getOperand(0), *Op1 = I.getOperand(1);
+  auto *II = dyn_cast<IntrinsicInst>(Op1);
+  if (!II || II->getIntrinsicID() != Intrinsic::sqrt || !II->hasOneUse() ||
+      !I.hasAllowReassoc() || !I.hasAllowReciprocal())
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1712]: Do we need `reassoc`?

Comment (zjaffal) [reply llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1712]: my assumption was following  the implementation of `foldFDivPowDivisor` or at least the div inside the sqrt should have that flag enabled?

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp [hunk 3]
```diff
@@ -1701,6 +1701,31 @@ static Instruction *foldFDivPowDivisor(BinaryOperator &I,
...
 }
 
+/// Convert div to mul if we have an sqrt divisor iff sqrt's operand is a fdiv
+/// instruction.
+static Instruction *foldFDivSqrtDivisor(BinaryOperator &I,
+                                        InstCombiner::BuilderTy &Builder) {
+  // X / sqrt(Y / Z) -->  X * sqrt(Z / Y)
+  Value *Op0 = I.getOperand(0), *Op1 = I.getOperand(1);
+  auto *II = dyn_cast<IntrinsicInst>(Op1);
+  if (!II || II->getIntrinsicID() != Intrinsic::sqrt || !II->hasOneUse() ||
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1711]: Would be cleaner to use PatternMatch. Something like

if (match(I, m_Sqrt(m_OneUse(m_FDiv(m_Value(Op0), m_Value(Op1))) {

}

File: llvm/test/Transforms/InstCombine/fdiv-sqrt.ll [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+; CHECK-NEXT:    call void @use(double [[DIV]])
+; CHECK-NEXT:    [[SQRT:%.*]] = call fast double @llvm.sqrt.f64(double [[DIV]])
+; CHECK-NEXT:    [[DIV1:%.*]] = fdiv fast double [[X:%.*]], [[SQRT]]
+; CHECK-NEXT:    ret double [[DIV1]]
+;
+entry:
+  %div = fdiv fast double %y, %z
+  call void @use(double %div)
+  %sqrt = call fast double @llvm.sqrt.f64(double %div)
+  %div1 = fdiv fast double %x, %sqrt
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/fdiv-sqrt.ll:65]: Should use reduced set of flags, and have some cases where flags are missing from individual instructions

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp (hunk 3 above)
Comment (zjaffal) [reply llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1711]: I agree but then we won't be able to check if `sqrt` has one use or `fdiv` has the necessary flags

File: llvm/test/Transforms/InstCombine/fdiv-sqrt.ll (hunk 4 above)
Comment (zjaffal) [reply llvm/test/Transforms/InstCombine/fdiv-sqrt.ll:65]: > Should use reduced set of flags
Is it satisfactory to assume only `reassoc arcp` are enabled? 

> have some cases where flags are missing from individual instructions
Will do thanks for the suggestion

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp (hunk 2 above)
Comment (jcranmer-intel) [reply llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1712]: Using `x.recip()` as shorthand for `(1.0 / x)`, `arcp` allows us to convert between `a / b` and `a * b.recip()` freely, and arguably `(a / b).recip()` to `(b / a)` (at the very least, gcc will do this with just `-freciprocal-math`).

We can convert `x / sqrt(y / z)` to `x * sqrt(y / z).recip()` with just `arcp`, and we can also convert `x * sqrt((y / z).recip())` to `x * sqrt(z / y)` with just `arcp` as well, but the question is which flags are necessary to convert `sqrt(a).recip()` to `sqrt(a.recip())`. I think it stretches `arcp` too far to allow this kind of permutation of `recip`. `reassoc` isn't entirely the right flag either; what we probably want is a more generic "allow algebraic identity" flag, but we're already using `reassoc` for that purpose elsewhere, so we might as well use it here as well.

Comment (zjaffal) [reply llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1712]: okay so from my understanding then the pattern we are looking for is the following 

```
%div = fdiv arcp double %y, %z
%sqrt = call reassoc double @llvm.sqrt.f64(double %div)
%div2 = fdiv arcp double %x, %sqrt
```

Comment (jcranmer-intel) [reply llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1712]: Not quite, you'd need the `reassoc` on the first `fdiv` as well, and at that point, it's probably best to just require that all the operations have `reassoc` and `arcp`.

**Comment by zjaffal** [comment]
ping

**Comment by zjaffal** [comment]
ping

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp [hunk 5]
```diff
@@ -1701,6 +1701,33 @@ static Instruction *foldFDivPowDivisor(BinaryOperator &I,
...
+    return nullptr;
+  Value *Op0 = I.getOperand(0), *Op1 = I.getOperand(1);
+  auto *II = dyn_cast<IntrinsicInst>(Op1);
+  if (!II || II->getIntrinsicID() != Intrinsic::sqrt || !II->hasOneUse() ||
+      !II->hasAllowReassoc() || !II->hasAllowReciprocal())
+    return nullptr;
+
+  Value *Y, *Z;
+  auto *DivOp = dyn_cast<Instruction>(II->getOperand(0));
+  if (!DivOp || !DivOp->hasOneUse() || !DivOp->hasAllowReassoc() ||
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1719]: Move hasOneUse to the end?

File: llvm/test/Transforms/InstCombine/fdiv-sqrt.ll [hunk 6]
```diff
@@ -0,0 +1,175 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py
+; RUN: opt -S -passes=instcombine < %s | FileCheck %s
+
+declare double @llvm.sqrt.f64(double)
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/fdiv-sqrt.ll:4]: Pre-submit baseline tests?

**Review by arsenm — APPROVED** [review]

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp (hunk 5 above)
Comment (zjaffal) [reply llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1719]: Do you mean just pushing the argument to be evaluated at the end? 
like this?
```suggestion
  if (!DivOp || !DivOp->hasAllowReassoc() || !I.hasAllowResprocal() || !DivOp->hasOneUse())
```

File: llvm/test/Transforms/InstCombine/fdiv-sqrt.ll (hunk 6 above)
Comment (zjaffal) [reply llvm/test/Transforms/InstCombine/fdiv-sqrt.ll:4]: In a separate PR?

File: llvm/test/Transforms/InstCombine/fdiv-sqrt.ll (hunk 4 above)
Comment (arsenm) [reply llvm/test/Transforms/InstCombine/fdiv-sqrt.ll:155]: Yes, I think so

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp (hunk 5 above)
Comment (arsenm) [reply llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1719]: Yes

File: llvm/test/Transforms/InstCombine/fdiv-sqrt.ll (hunk 6 above)
Comment (arsenm) [reply llvm/test/Transforms/InstCombine/fdiv-sqrt.ll:4]: Yes, or directly push

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp [hunk 7]
```diff
@@ -1701,6 +1701,33 @@ static Instruction *foldFDivPowDivisor(BinaryOperator &I,
...
+  Value *Op0 = I.getOperand(0), *Op1 = I.getOperand(1);
+  auto *II = dyn_cast<IntrinsicInst>(Op1);
+  if (!II || II->getIntrinsicID() != Intrinsic::sqrt || !II->hasOneUse() ||
+      !II->hasAllowReassoc() || !II->hasAllowReciprocal())
+    return nullptr;
+
+  Value *Y, *Z;
+  auto *DivOp = dyn_cast<Instruction>(II->getOperand(0));
+  if (!DivOp || !DivOp->hasAllowReassoc() || !I.hasAllowReciprocal() || !DivOp->hasOneUse())
+      !I.hasAllowReciprocal())
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1720]:
```suggestion
  if (!DivOp || !DivOp->hasAllowReassoc() || !I.hasAllowReciprocal() ||
       !DivOp->hasOneUse())
```

**Comment by zjaffal** [comment]
@arsenm thanks for your help i will push the test commits directly to main and then rebase this branch and merge

**Comment by zjaffal** [comment]
@arsenm i forced push the branch after cherry picking the tests to main now every check passes
shall I merge it

**Review by arsenm — APPROVED** [review]

**Pull request merged**